### PR TITLE
Upgrade Pester tests in test\powershell\Host folder to dashed v4 syntax

### PR DESCRIPTION
**Pull request by bergmeister**
## PR Summary

Fixes a subset of #6245
Use new dashed Pester v4 operators for tests in `test\powershell\Host`

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [NA] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by iSazonov** [comment]
@bergmeister Please clarify about "-Not -Throw is deprecated"

/cc @it-praktyk @nohwnd

File: test/powershell/Host/PSVersionTable.Tests.ps1 [hunk 1]
```diff
@@ -25,51 +25,51 @@ Describe "PSVersionTable" -Tags "CI" {
...
-       $PSVersionTable.ContainsKey("PSVersion")                 | Should Be True
-       $PSVersionTable.ContainsKey("PSEdition")                 | Should Be True
-       $PSVersionTable.ContainsKey("WSManStackVersion")         | Should Be True
-       $PSVersionTable.ContainsKey("SerializationVersion")      | Should Be True
-       $PSVersionTable.ContainsKey("PSCompatibleVersions")      | Should Be True
-       $PSVersionTable.ContainsKey("PSRemotingProtocolVersion") | Should Be True
-       $PSVersionTable.ContainsKey("GitCommitId")               | Should Be True
-       $PSVersionTable.ContainsKey("Platform")                  | Should Be True
-       $PSVersionTable.ContainsKey("OS")                        | Should Be True
+       $PSVersionTable.ContainsKey("PSVersion")                 | Should -Be True
```
Comment (iSazonov) [inline test/powershell/Host/PSVersionTable.Tests.ps1]: We can use `-BeTrue`.

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 2]
```diff
@@ -115,20 +115,17 @@ Describe "ConsoleHost unit tests" -tags "Feature" {
...
             }
         }
 
-        It "Verify Validate Output Format As Text Explicitly Child Single Shell should works" {
-            {
-                $a="blahblah"
-                $a | & $powershell -noprofile -out text -com { $input }
-            } | Should Not Throw
+        It "Verify Validate Output Format As Text Explicitly Child Single Shell doess not throw" {
+            "blahblah" | & $powershell -noprofile -out text -com { $input }
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: Typo in title.

**Review by iSazonov — CHANGES_REQUESTED** [review]

**Comment by bergmeister** [comment]
@iSazonov  See [here](https://github.com/pester/Pester/releases/tag/4.2.0) (bottom section about breaking changes) about why `Should -Not -Throw` should not be used any more. I agree with your other suggestions.

**Comment by iSazonov** [comment]
If you means:
>Should -Not -Throw will fail if any exception is thrown. The filters applied on Should -Not -Throw do not apply, because when an exception is thrown, it automatically means that the test should not pass.

I don't see where said that it is "deprecated". My understanding is that with the change we get expected behavior "Should -Not -Throw _will fail_ if any exception is thrown."

**Comment by bergmeister** [comment]
I would like to hear first what the Pester guys suggest. I read it first the sentence `Should -Not -Throw will fail if any exception is thrown.` as that the operator itself will fail but you might be right that they changed it to make it work. But in general I do not understand why such an assertion is needed at all. Does a Pester test not fail automatically if an terminating or non-terminating error occurs?

**Comment by iSazonov** [comment]
Pester `Should` like C# [Fact] in Unit tests - we should _explicitly_ see that we expect. Also if we remove `Should` we (can) get noisy output and we have to suppress it by redirection to null.

**Comment by nohwnd** [comment]
`Should -Not -Throw` is not deprecated. Only the behavior changed. Before, `Should -Not -Throw`was a negation of `Should -Throw`, which allowed it to catch an exception, and still pass the test. So I corrected the behavior to fail if any exception is thrown. And now I think I should make the assertion fail, if `-Not` and any other filter is used.

The change of behavior no longer allows for writing tests that say that "code should not throw a particular exception, but any other exception is fine". This is a sacrifice I am willing to make, as this kind of test does not make much sense, and is easy to replicate via `Should -Throw`. 

Personally I would remove `Should -Not -Throw` from the assertion library completely, if it was easily possible. Because as with any other .NET test framework, any exception thrown within a test fails the test. So `Should -Not -Throw` is implied on every line of code that we write, and there is no need to explicitly call it out. In fact it only help us to write incomplete tests, because it looks like we have an assertion, while we should be testing for a result or an effect instead.

@iSazonov I am not sure, what you mean by your last remark about `Should` and `Fact`. Should is an assertion, but Fact is an attribute that mark a test in xUnit, did you mean Assert?. Anyways removing Should -Not -Throw, shouldn't be a problem.

**Comment by iSazonov** [comment]
@nohwnd Thanks for confirmation.
I means that "Should" brings useful information - we _explicitly_ see the expected result. If this is not present, the `It` block looks like useless - we can not know that the expected result of code execution should be an exception - the first thought is that we skip anything.

**Comment by nohwnd** [comment]
@iSazonov 
> we explicitly see the expected result.

I cannot agree with that. That code does not throw is not a good result of a test. A test is supposed to validate that something useful happened. A function can be useful by either returning the correct result, or producing the correct side effect. (eg. adding 1+1 and making sure you got 2, or deleting a file and making sure it's deleted).  

A function that does not return any value, and does not have any side effects is not useful. (but feel free to prove me wrong :) )

Coming back to our useful functions we can see that pre-requisites for successfully returning the correct value, or producing the correct side-effect is that the code does not throw. So a test that only validates that you did not throw, is just a subset of a more useful test that would test the resulting value, or the resulting side-effect.

( If this does not make a sense, you can view it from TDD point of view. A test is a mean that allows you to force your code to have the desired behavior. By writing a test that validates that your function returns 2 for input 1 and 1 you are forcing the function to implement some behavior, because by default it does not return anything. BUT by expecting the code to not throw, you are not forcing anything new, because empty functions do not throw by default. )

**Comment by iSazonov** [comment]
This logic is perfectly correct in the general case. We would have to do all our tests that way. But we have _edge_ cases where we are forced to use this "Should -Not -Throw". For example, in this PR we say that modified psversiontable can be formatted, but we do not check the exact result - such a test would be unduly complicated, so we just check that there is no exception - then we _trust_ formatting system tests that exist separately. There are other examples when there is output to the console, user interaction, changing system settings (Set-Date). So "Should -Not -Throw" helps us to simplify tests if it is justified or impossible otherwise.

**Comment by bergmeister** [comment]
So what is the outcome? Do I need to revert the removal of `Should -Not -Throw`?

**Comment by iSazonov** [comment]
@bergmeister Yes, please revert - we'll use this until we make a decision to replace such edge cases with full positive tests.

**Review by iSazonov — APPROVED** [review]

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by bergmeister** [comment]
@daxian-dbw Please tick the corresponding checkbox in the referenced issue 6245
